### PR TITLE
feat(airbyte-cdk) Align BackoffStrategy interfaces to take attempt_count as a full-fled…

### PR DESCRIPTION
**Pull request by maxi297**
…ge parameter

## What
While doing some development in the backoff strategies, I got a very weird surprise: some backoff strategies ([here](https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/error_handlers/backoff_strategies/wait_time_from_header_backoff_strategy.py#L39) and [here](https://github.com/airbytehq/airbyte/blob/1d24db9b29eacc92d6d20d986c7160d3e3a63d52/airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/error_handlers/backoff_strategies/wait_until_time_from_header_backoff_strategy.py#L42)) implements `BackoffStrategy` while not matching the interface as they have an extra parameter `attempt_count`. Others have an interface that match but will fail if it is not provided ([here](https://github.com/airbytehq/airbyte/blob/1d24db9b29eacc92d6d20d986c7160d3e3a63d52/airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/error_handlers/backoff_strategies/exponential_backoff_strategy.py#L42-L44)). 

This shows that the backoff strategies in the declarative package can’t be used as generic backoff strategies as if a caller does not provide `attempt_count` as a keyword parameter (which is not explicit as part of the interface), the connector will crash. 

Here, we are trying to achieve that it is possible to swap any implementation of `BackoffStrategy` without having to change anything.

## How
Adding `attempt_count` as a mandatory parameter. The migration notes describe the implication of that

## User Impact
Some tests will break so we will socialize this change and make it clear that is someone has the error `backoff_time() missing 1 required positional argument: 'attempt_count'`, they now need to pass the `attempt_count` parameter.

## Can this PR be safely reverted and rolled back?
- [X] YES 💚
- [ ] NO ❌

### CODE REVIEW

**Review by pnilan — APPROVED** [review]
LGTM

File: airbyte-cdk/python/cdk-migrations.md [hunk 1]
```diff
@@ -1,5 +1,19 @@
...
+
+Main impact: This change is mostly internal but we spotted a couple of tests that expect `backoff_time` to not have the `attempt_count` parameter so these tests would fail ([example](https://github.com/airbytehq/airbyte/blob/c9f45a0b85735f58102fcd78385f6f673e731aa6/airbyte-integrations/connectors/source-github/unit_tests/test_stream.py#L99)).
+
+This change should not impact the following classes even though they have a different interface as they accept `kwargs` and  `attempt_count` is currently passed as a keyword argument within the CDK. However, once there is a CDK change where `backoff_time` is called not as a keyword argument, they will fail: 
+* Zendesk Support: ZendeskSupportBackoffStrategy (this one will be updated shortly after as it is used for CI to validate CDK changes)
+* Klaviyo: KlaviyoBackoffStrategy (the logic has been generified so we will remove this custom component shortly after this update)
+* GitHub: GithubStreamABCBackoffStrategy and ContributorActivityBackoffStrategy
+* Airtable: AirtableBackoffStrategy
+* Slack: SlackBackoffStrategy
```
Comment (brianjlai) [inline airbyte-cdk/python/cdk-migrations.md:13]: thanks for auditing the existing connectors in our repo

**Review by brianjlai — APPROVED** [review]
lgtm!

**Pull request merged**